// hydrate-mongodb/query/queryKind.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var QueryKind;
(function (QueryKind) {
    QueryKind[QueryKind["FindAll"] = 0] = "FindAll";
    QueryKind[QueryKind["FindEach"] = 1] = "FindEach";
    QueryKind[QueryKind["FindEachSeries"] = 2] = "FindEachSeries";
    QueryKind[QueryKind["FindOne"] = 3] = "FindOne";
    QueryKind[QueryKind["FindOneById"] = 4] = "FindOneById";
    QueryKind[QueryKind["FindOneAndRemove"] = 5] = "FindOneAndRemove";
    QueryKind[QueryKind["FindOneAndUpdate"] = 6] = "FindOneAndUpdate";
    QueryKind[QueryKind["FindOneAndUpsert"] = 7] = "FindOneAndUpsert";
    QueryKind[QueryKind["FindCursor"] = 8] = "FindCursor";
    QueryKind[QueryKind["RemoveAll"] = 9] = "RemoveAll";
    QueryKind[QueryKind["RemoveOne"] = 10] = "RemoveOne";
    QueryKind[QueryKind["UpdateAll"] = 11] = "UpdateAll";
    QueryKind[QueryKind["UpdateOne"] = 12] = "UpdateOne";
    QueryKind[QueryKind["Upsert"] = 13] = "Upsert";
    QueryKind[QueryKind["Distinct"] = 14] = "Distinct";
    QueryKind[QueryKind["Count"] = 15] = "Count";
})(QueryKind = exports.QueryKind || (exports.QueryKind = {}));
