// hydrate-mongodb/mapping/readContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var mappingError_1 = require("./mappingError");
/**
 * @hidden
 */
var ReadContext = /** @class */ (function () {
    function ReadContext(session) {
        this.session = session;
        /**
         * The current path.
         */
        this.path = "";
    }
    /**
     * Adds an error to the context.
     * @param message The error message.
     * @param path Optional. The current path if different than what's in the context.
     */
    ReadContext.prototype.addError = function (message, path) {
        if (!this.errors) {
            this.errors = [];
            this.hasErrors = true;
        }
        this.errors.push({
            message: message,
            path: path || this.path
        });
    };
    ReadContext.prototype.addFetch = function (path) {
        if (!this.fetches) {
            this.fetches = [];
        }
        this.fetches.push(path);
    };
    /**
     * Gets a string summarizing all errors in the context.
     */
    ReadContext.prototype.getErrorMessage = function () {
        return mappingError_1.createErrorMessage(this.errors);
    };
    return ReadContext;
}());
exports.ReadContext = ReadContext;
