// hydrate-mongodb/mapping/writeContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var mappingError_1 = require("./mappingError");
/**
 * @hidden
 */
var WriteContext = /** @class */ (function () {
    /**
     * Constructs a Write Context.
     * @param path The current path.
     */
    function WriteContext(path) {
        if (path === void 0) { path = ""; }
        this.path = path;
        /**
         * A list of errors that occurred during serialization.
         */
        this.errors = [];
        /**
         * Listed of objects that have been visited to prevent infinite loops.
         */
        this.visited = [];
    }
    /**
     * Adds an error to the context.
     * @param message The error message.
     * @param path Optional. The current path if different than what's in the context.
     */
    WriteContext.prototype.addError = function (message, path) {
        this.errors.push({
            message: message,
            path: path || this.path
        });
        this.hasErrors = true;
    };
    /**
     * Gets a string summarizing all errors in the context.
     */
    WriteContext.prototype.getErrorMessage = function () {
        return mappingError_1.createErrorMessage(this.errors);
    };
    return WriteContext;
}());
exports.WriteContext = WriteContext;
