// hydrate-mongodb/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var reference_1 = require("./reference");
var namingStrategies_1 = require("./config/namingStrategies");
exports.NamingStrategies = namingStrategies_1.NamingStrategies;
var persistenceError_1 = require("./persistenceError");
exports.PersistenceError = persistenceError_1.PersistenceError;
exports.EntityNotFoundError = persistenceError_1.EntityNotFoundError;
var configuration_1 = require("./config/configuration");
exports.Configuration = configuration_1.Configuration;
var objectIdGenerator_1 = require("./config/objectIdGenerator");
exports.ObjectIdGenerator = objectIdGenerator_1.ObjectIdGenerator;
var annotationMappingProvider_1 = require("./mapping/providers/annotationMappingProvider");
exports.AnnotationMappingProvider = annotationMappingProvider_1.AnnotationMappingProvider;
__export(require("./mapping/providers/decorators"));
function areEqual(value1, value2) {
    return reference_1.Reference.areEqual(value1, value2);
}
exports.areEqual = areEqual;
