// hydrate-mongodb/mapping/providers/decorators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var reflect_helper_1 = require("reflect-helper");
var annotations_1 = require("./annotations");
exports.Entity = reflect_helper_1.makeDecorator(annotations_1.EntityAnnotation);
exports.Embeddable = reflect_helper_1.makeDecorator(annotations_1.EmbeddableAnnotation);
exports.Converter = reflect_helper_1.makeDecorator(annotations_1.ConverterAnnotation);
exports.Collection = reflect_helper_1.makeDecorator(annotations_1.CollectionAnnotation);
exports.Index = reflect_helper_1.makeDecorator(annotations_1.IndexAnnotation);
exports.VersionField = reflect_helper_1.makeDecorator(annotations_1.VersionFieldAnnotation);
exports.Versioned = reflect_helper_1.makeDecorator(annotations_1.VersionedAnnotation);
exports.ChangeTracking = reflect_helper_1.makeDecorator(annotations_1.ChangeTrackingAnnotation);
exports.DiscriminatorField = reflect_helper_1.makeDecorator(annotations_1.DiscriminatorFieldAnnotation);
exports.DiscriminatorValue = reflect_helper_1.makeDecorator(annotations_1.DiscriminatorValueAnnotation);
exports.Field = reflect_helper_1.makeDecorator(annotations_1.FieldAnnotation);
exports.Parent = reflect_helper_1.makeDecorator(annotations_1.ParentAnnotation);
exports.Enumerated = reflect_helper_1.makeDecorator(annotations_1.EnumeratedAnnotation);
exports.Cascade = reflect_helper_1.makeDecorator(annotations_1.CascadeAnnotation);
exports.Fetch = reflect_helper_1.makeDecorator(annotations_1.FetchAnnotation);
exports.InverseOf = reflect_helper_1.makeDecorator(annotations_1.InverseOfAnnotation);
exports.Type = reflect_helper_1.makeDecorator(annotations_1.TypeAnnotation);
exports.ElementType = reflect_helper_1.makeDecorator(annotations_1.ElementTypeAnnotation);
exports.MapKey = reflect_helper_1.makeDecorator(annotations_1.MapKeyAnnotation);
exports.Immutable = reflect_helper_1.makeDecorator(annotations_1.ImmutableAnnotation);
exports.Transient = reflect_helper_1.makeDecorator(annotations_1.TransientAnnotation);
exports.Identity = reflect_helper_1.makeDecorator(annotations_1.IdentityAnnotation);
